### DISCUSSION TITLE
Triggering "Action-Widgets" without a user input

### DISCUSSION THREAD

**Discussion by msameen99** [TiddlyWiki/TiddlyWiki5#6380] [Q&A] [answered]
Is there a way to trigger "Action Widget" (ex. action-createtiddler) without a user input? either from a JavaScript Macro or WikiText Macro?

In other words, what is my options to "create tiddlers" without user inputs from within my Macros (WikiText or JavaScript)?

**Comment by pmario** [comment]
I did create a `trigger` widget, that triggers actions as soon as it is rendered. ... The main problem, why I didn't promote it, is, that it is possible to create endless loops, that can brick your wiki, if done wrong. 

See: https://wikilabs.github.io/editions/trigger/ ... There is not much docs except the code of the test framework itself. ... In the "demo" wiki I did still use buttons to trigger "next" and "previous" actions. ... BUT it would be possible to create something that "runs" automatically, since an active action can trigger the next one by opening the tiddler, which contains that action.

**Comment by pmario** [comment]
Eric Shulman has a similar widget, that can be tirggered with a timer. See https://talk.tiddlywiki.org/t/tw5-export-tiddler-automatically/1929/3?u=pmario

**Comment by saqimtiaz** [comment]
@msameen99 It would be helpful if you described the broader context of what you are trying to do. If this is in relation to the "heavy calculations" that you mentioned in another thread, what is the source of the data for the calculations? When do the calculations need to take place: on wiki load, when a tiddler is opened or in response to another trigger?

**Reply by msameen99** [reply]
@saqimtiaz 

1- Yes, this related to the "heavy calculations" thread
2- The source is:
- "currentTiddler" Title.
- "ALL other Tiddlers" to collect the existed relational with 'currentTiddler".
- External API calls using TW5's data as parameters.

3- Calculations will start whenever a tiddler is opened/edited/rendered, and the result will displayed in a "custom view template" inside the currentTiddler (think as a Footer for each Tiddler).

I did a WikiText implementation (not for API) and use it for more than a year now, it's great, but its "Information Model" is limited and the code is messy and not intuitive, also whenever I open more Tiddlers, TW5 become slower and slower (my wiki has ~ 4000 Tiddlers).

So, I'm planing to do the next version (v4), to improve both the "Information Model" and the Code.
For the code part I'll use JavaScript to Minimize the calculation time as possible and to handle more complex logic (WikiText is GREAT but has its limits).

To get a taste of it, please check the following tiddler example (all other tiddlers have a similar simple design), in the edit mode they look like :
![image](https://user-images.githubusercontent.com/16881960/147989877-71641f7e-1bd6-49fe-a949-74496ef56896.png)

and here is how they rendered:
![image](https://user-images.githubusercontent.com/16881960/147990387-fad73653-9de1-41ee-8217-8411d5b3d828.png)

The Calculation dynamically get all currentTiddler's existing relations, and show ONLY the Horizontal->Vertical taps for them, also listing all Tiddlers Title that involved in each relation.
here is another Example:
![image](https://user-images.githubusercontent.com/16881960/147991264-ffd78c43-e310-4f12-8fba-f67c7df383e8.png)

**Comment by Jermolene** [comment, marked as answer]
> Is there a way to trigger "Action Widget" (ex. action-createtiddler) without a user input? either from a JavaScript Macro or WikiText Macro?

Action widgets can only be triggered by events, particularly user interface events like button clicks, but also timers and scroll events.

JavaScript macros must not have side effects, and so must not trigger action widgets.

> In other words, what is my options to "create tiddlers" without user inputs from within my Macros (WikiText or JavaScript)?

A common pattern for calling a JS function that returns a large amount of structured data is to use a widget message to trigger the function, passing parameters as required. The function should return its results by creating as many interlinked tiddlers as it needs.

The plugin library uses this pattern: the `tm-load-plugin-library` triggers an action and takes a parameter specifying the tiddlers that should receive the results.

See: https://tiddlywiki.com/#WidgetMessage%3A%20tm-load-plugin-library

**Reply by msameen99** [reply]
Thanks a lot Jeremy for your detailed reply, I really appreciate it , 
I'll see how the "widget message" can help, kindly check my above reply to Saq to have an idea about what i'm trying to do.
